diff -Nur instead-2.5.0/debian/compat instead-2.5.0-maemo/debian/compat
--- instead-2.5.0/debian/compat	2017-03-11 10:34:42.000000000 +0300
+++ instead-2.5.0-maemo/debian/compat	2017-03-11 13:04:27.792981000 +0300
@@ -1 +1 @@
-7
+4
diff -Nur instead-2.5.0/debian/control instead-2.5.0-maemo/debian/control
--- instead-2.5.0/debian/control	2017-03-11 10:34:42.000000000 +0300
+++ instead-2.5.0-maemo/debian/control	2017-03-11 13:04:27.793261000 +0300
@@ -1,8 +1,8 @@
 Source: instead
-Section: games
+Section: user/games
 Priority: optional
 Maintainer: Peter Kosyh <p.kosyh@gmail.com>
-Build-Depends: debhelper (>= 7), pkg-config, liblua5.1-dev, libsdl1.2-dev, libsdl-ttf2.0-dev, libsdl-image1.2-dev, libsdl-mixer1.2-dev, zlib1g-dev
+Build-Depends: debhelper (>= 5), pkg-config, liblua5.1-dev, libsdl1.2-dev, libsdl-ttf2.0-dev, libsdl-image1.2-dev, libsdl-mixer1.2-dev, zlib1g-dev
 Standards-Version: 3.8.3
 Homepage: http://instead.sourceforge.net
 
@@ -11,4 +11,3 @@
 Depends: ${shlibs:Depends}, ${misc:Depends}
 Description: simple text adventures/visual novels engine
  Visual novell/text quest-like engine.
-
diff -Nur instead-2.5.0/debian/rules instead-2.5.0-maemo/debian/rules
--- instead-2.5.0/debian/rules	2017-03-11 10:34:42.000000000 +0300
+++ instead-2.5.0-maemo/debian/rules	2017-03-11 13:04:27.793529000 +0300
@@ -61,7 +61,7 @@
 install: build
 	dh_testdir
 	dh_testroot
-	dh_prep
+#	dh_prep
 	dh_installdirs
 
 	# Add here commands to install the package into debian/instead.
diff -Nur instead-2.5.0/desktop/instead.desktop.in instead-2.5.0-maemo/desktop/instead.desktop.in
--- instead-2.5.0/desktop/instead.desktop.in	2017-03-11 10:34:42.000000000 +0300
+++ instead-2.5.0-maemo/desktop/instead.desktop.in	2017-03-11 13:04:27.793840000 +0300
@@ -13,4 +13,8 @@
 Exec=@BIN@/sdl-instead
 Icon=sdl_instead
 Terminal=false
+X-Window-Icon=sdl_instead
+X-Window-Icon-Dimmed=sdl_instead
+X-HildonDesk-ShowInToolbar=true
+StartupWMClass=instead
 Categories=Game;LogicGame;
diff -Nur instead-2.5.0/desktop/Makefile instead-2.5.0-maemo/desktop/Makefile
--- instead-2.5.0/desktop/Makefile	2017-03-11 10:34:42.000000000 +0300
+++ instead-2.5.0-maemo/desktop/Makefile	2017-03-11 13:04:27.794105000 +0300
@@ -10,9 +10,9 @@
 	cat instead.desktop.in | sed -e "s|@BIN@|$(BIN)|g" > instead.desktop
 
 install:
-	$(INSTALLD) $(DESTDIR)$(PREFIX)/share/applications
-	$(INSTALL) instead.desktop $(DESTDIR)$(PREFIX)/share/applications/instead.desktop
+	$(INSTALLD) $(DESTDIR)$(PREFIX)/share/applications/hildon
+	$(INSTALL) instead.desktop $(DESTDIR)$(PREFIX)/share/applications/hildon/instead.desktop
 
 uninstall:
-	$(RM) $(DESTDIR)$(PREFIX)/share/applications/instead.desktop
+	$(RM) $(DESTDIR)$(PREFIX)/share/applications/hildon/instead.desktop
 
diff -Nur instead-2.5.0/patches/maemo.patch instead-2.5.0-maemo/patches/maemo.patch
diff -Nur instead-2.5.0/Rules.make.system instead-2.5.0-maemo/Rules.make.system
--- instead-2.5.0/Rules.make.system	2017-03-11 10:34:42.000000000 +0300
+++ instead-2.5.0-maemo/Rules.make.system	2017-03-11 13:04:27.794345000 +0300
@@ -28,7 +28,7 @@
 SDL_CFLAGS=$(shell sdl-config --cflags)
 SDL_LFLAGS=$(shell sdl-config --libs) -lSDL_ttf -lSDL_mixer -lSDL_image -lm
 
-CFLAGS	+= -Wall -Dunix -D_USE_UNPACK # -D_SDL_MOD_BUG
+CFLAGS	+= -Wall -Dunix -D_USE_UNPACK -DMAEMO # -D_SDL_MOD_BUG
 
 INSTALLD=install -d -m 0755
 INSTALLB=install -m 0755
diff -Nur instead-2.5.0/src/instead/util.h instead-2.5.0-maemo/src/instead/util.h
--- instead-2.5.0/src/instead/util.h	2017-03-11 10:34:44.000000000 +0300
+++ instead-2.5.0-maemo/src/instead/util.h	2017-03-11 13:04:46.785519000 +0300
@@ -52,7 +52,7 @@
 #define FREE(v) do { if ((v)) free((v)); v = NULL; } while(0)
 
 #ifdef _USE_SDL
- #include <SDL_config.h>
+/* #include <SDL_config.h> */
  #include <SDL_mutex.h>
  #ifdef _USE_SDL_ICONV
   #include <SDL_stdinc.h>
diff -Nur instead-2.5.0/src/sound.c instead-2.5.0-maemo/src/sound.c
--- instead-2.5.0/src/sound.c	2017-03-11 10:34:44.000000000 +0300
+++ instead-2.5.0-maemo/src/sound.c	2017-03-11 14:28:39.065602000 +0300
@@ -27,7 +27,9 @@
 
 #include <SDL.h>
 #include <SDL_mixer.h>
-
+#define SDL_memcpy memcpy
+#define SDL_calloc calloc
+#define SDL_free free
 #ifdef S60
 int audio_rate = 11025;
 #else
diff -Nur instead-2.5.0/themes/default/theme.ini instead-2.5.0-maemo/themes/default/theme.ini
--- instead-2.5.0/themes/default/theme.ini	2017-03-11 10:34:45.000000000 +0300
+++ instead-2.5.0-maemo/themes/default/theme.ini	2017-03-11 13:04:27.794651000 +0300
@@ -5,7 +5,7 @@
 ; $Name(it): Predefinito$
 
 scr.w = 800
-scr.h = 600
+scr.h = 480
 scr.gfx.scalable = 1
 
 scr.gfx.h = -1
@@ -23,7 +23,7 @@
 win.x = 48
 win.y = 8
 win.w = 500
-win.h = 568
+win.h = 448
 
 win.fnt.name = {sans,sans-b,sans-i,sans-bi}.ttf
 win.fnt.size = 16
@@ -43,7 +43,8 @@
 inv.x = 620
 inv.y = 8
 inv.w = 144
-inv.h = 564
+inv.h = 448
+
 
 inv.fnt.name = {sans,sans-b,sans-i,sans-bi}.ttf
 inv.fnt.size = 16
@@ -71,8 +72,8 @@
 menu.fnt.height = 1.0
 menu.gfx.button = menu.png
 menu.button.x = 776
-menu.button.y = 576
 
+menu.button.y = 456
 
 snd.click = click.wav ; click.ogg
 
diff -Nur instead-2.5.0/themes/wide/theme.ini instead-2.5.0-maemo/themes/wide/theme.ini
--- instead-2.5.0/themes/wide/theme.ini	2017-03-11 10:34:46.000000000 +0300
+++ instead-2.5.0-maemo/themes/wide/theme.ini	2017-03-11 13:04:27.794891000 +0300
@@ -3,10 +3,3 @@
 ; $Name(uk):Широка$
 ; $Name(es):Amplio$
 ; $Name(it):Ampio$
-
-scr.w = 800
-scr.h = 480
-win.h = 448
-inv.h = 448
-
-menu.button.y = 456
