diff -Nur instead-1.4.1/debian/compat instead-1.4.1-maemo/debian/compat
--- instead-1.4.1/debian/compat	2011-05-13 11:40:08.000000000 +0400
+++ instead-1.4.1-maemo/debian/compat	2011-05-11 07:45:43.357935817 +0400
@@ -1 +1 @@
-7
+4
diff -Nur instead-1.4.1/debian/control instead-1.4.1-maemo/debian/control
--- instead-1.4.1/debian/control	2011-05-13 11:40:08.000000000 +0400
+++ instead-1.4.1-maemo/debian/control	2011-05-11 07:45:43.357935817 +0400
@@ -1,8 +1,8 @@
 Source: instead
-Section: games
+Section: user/games
 Priority: optional
 Maintainer: Peter Kosyh <p.kosyh@gmail.com>
-Build-Depends: debhelper (>= 7), pkg-config, liblua5.1-dev, libsdl1.2-dev, libsdl-ttf2.0-dev, libsdl-image1.2-dev, libsdl-mixer1.2-dev, zlib1g-dev
+Build-Depends: debhelper (>= 5), pkg-config, liblua5.1-dev, libsdl1.2-dev, libsdl-ttf2.0-dev, libsdl-image1.2-dev, libsdl-mixer1.2-dev, zlib1g-dev
 Standards-Version: 3.8.3
 Homepage: http://instead.googlecode.com
 
@@ -11,4 +11,3 @@
 Depends: ${shlibs:Depends}, ${misc:Depends}
 Description: simply text adventures/visual novels engine and game
  Visual novell/text quest-like game in Russian with engine.
-
diff -Nur instead-1.4.1/debian/rules instead-1.4.1-maemo/debian/rules
--- instead-1.4.1/debian/rules	2011-05-13 11:40:08.000000000 +0400
+++ instead-1.4.1-maemo/debian/rules	2011-05-11 07:45:43.361932007 +0400
@@ -52,7 +52,7 @@
 install: build
 	dh_testdir
 	dh_testroot
-	dh_prep
+#	dh_prep
 	dh_installdirs
 
 	# Add here commands to install the package into debian/instead.
diff -Nur instead-1.4.1/desktop/instead.desktop.in instead-1.4.1-maemo/desktop/instead.desktop.in
--- instead-1.4.1/desktop/instead.desktop.in	2011-05-13 11:40:18.000000000 +0400
+++ instead-1.4.1-maemo/desktop/instead.desktop.in	2011-05-11 07:45:43.361932007 +0400
@@ -7,6 +7,10 @@
 Comment=Siple Text Adventures Interpreter
 Comment[ru]=Интерпретатор простых приключений
 Exec=@BIN/sdl-instead
-Icon=sdl_instead.png
+Icon=sdl_instead
 Terminal=false
+X-Window-Icon=sdl_instead
+X-Window-Icon-Dimmed=sdl_instead
+X-HildonDesk-ShowInToolbar=true
+StartupWMClass=instead
 Categories=Game;LogicGame;
diff -Nur instead-1.4.1/desktop/Makefile instead-1.4.1-maemo/desktop/Makefile
--- instead-1.4.1/desktop/Makefile	2011-05-13 11:40:18.000000000 +0400
+++ instead-1.4.1-maemo/desktop/Makefile	2011-05-11 07:45:43.361932007 +0400
@@ -10,9 +10,9 @@
 	cat instead.desktop.in | sed -e "s|@BIN|$(BIN)|g" > instead.desktop
 
 install:
-	$(INSTALLD) $(DESTDIR)$(PREFIX)/share/applications
-	$(INSTALL) instead.desktop $(DESTDIR)$(PREFIX)/share/applications/instead.desktop
+	$(INSTALLD) $(DESTDIR)$(PREFIX)/share/applications/hildon
+	$(INSTALL) instead.desktop $(DESTDIR)$(PREFIX)/share/applications/hildon/instead.desktop
 
 uninstall:
-	$(RM) $(DESTDIR)$(PREFIX)/share/applications/instead.desktop
+	$(RM) $(DESTDIR)$(PREFIX)/share/applications/hildon/instead.desktop
 
diff -Nur instead-1.4.1/Rules.make.system instead-1.4.1-maemo/Rules.make.system
--- instead-1.4.1/Rules.make.system	2011-05-13 11:40:07.000000000 +0400
+++ instead-1.4.1-maemo/Rules.make.system	2011-05-11 07:45:43.361932007 +0400
@@ -28,7 +28,7 @@
 SDL_CFLAGS=$(shell sdl-config --cflags)
 SDL_LFLAGS=$(shell sdl-config --libs) -lSDL_ttf -lSDL_mixer -lSDL_image
 
-CFLAGS	+= -Wall -Dunix -D_USE_UNPACK # -D_SDL_MOD_BUG
+CFLAGS	+= -Wall -Dunix -D_USE_UNPACK -DMAEMO # -D_SDL_MOD_BUG
 
 INSTALLD=install -d -m 0755
 INSTALLB=install -m 0755
diff -Nur instead-1.4.1/themes/default/theme.ini instead-1.4.1-maemo/themes/default/theme.ini
--- instead-1.4.1/themes/default/theme.ini	2011-05-13 11:40:10.000000000 +0400
+++ instead-1.4.1-maemo/themes/default/theme.ini	2011-05-11 07:49:15.577901900 +0400
@@ -4,7 +4,7 @@
 ; $Name(it): Predefinito$
 
 scr.w = 800
-scr.h = 600
+scr.h = 480
 scr.gfx.scalable = 1
 
 scr.gfx.h = -1
@@ -21,7 +21,7 @@
 win.x = 48
 win.y = 8
 win.w = 500
-win.h = 568
+win.h = 448
 
 win.fnt.name = {sans,sans-b,sans-i,sans-bi}.ttf
 win.fnt.size = 16
@@ -39,7 +39,8 @@
 inv.x = 620
 inv.y = 8
 inv.w = 144
-inv.h = 564
+inv.h = 448
+
 
 inv.fnt.name = {sans,sans-b,sans-i,sans-bi}.ttf
 inv.fnt.size = 16
@@ -67,8 +68,8 @@
 menu.fnt.height = 1.0
 menu.gfx.button = menu.png
 menu.button.x = 776
-menu.button.y = 576
 
+menu.button.y = 456
 
 snd.click = click.wav ; click.ogg
 
diff -Nur instead-1.4.1/themes/wide/theme.ini instead-1.4.1-maemo/themes/wide/theme.ini
--- instead-1.4.1/themes/wide/theme.ini	2011-05-13 11:40:11.000000000 +0400
+++ instead-1.4.1-maemo/themes/wide/theme.ini	2011-05-11 07:47:51.833957190 +0400
@@ -2,10 +2,3 @@
 ; $Name(ru):Широкая$
 ; $Name(es):Amplio$
 ; $Name(it):Ampio$
-
-scr.w = 800
-scr.h = 480
-win.h = 448
-inv.h = 448
-
-menu.button.y = 456
